### PR TITLE
patch(mattermost_alert.yaml): add mattermost_alert

### PR DESCRIPTION
**Pull request by marcoppenheimer**
#### [canonical/zookeeper-operator](https://github.com/canonical/zookeeper-operator/actions/runs/4734403462) has failing jobs
:x:   **[integration-password-rotation](https://github.com/canonical/zookeeper-operator/actions/runs/4734403462/jobs/8403373460)**
:x:   **[integration-scaling](https://github.com/canonical/zookeeper-operator/actions/runs/4734403462/jobs/8403373961)**

---

@marcoppenheimer @deusebio

## For Reviewers
Please leave ideas, nitpicks and thoughts on the alert post content and linking. There's a host of information available in the [Github Actions context](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context) that we can expose here, but I've tried to be conservative to avoid noise. Example data available that we can use:
```json
    {
      "id": 12838040994,
      "run_id": 4734403462,
      "workflow_name": "Tests",
      "head_branch": "fix/shared_etc_env",
      "run_url": "https://api.github.com/repos/canonical/zookeeper-operator/actions/runs/4734403462",
      "run_attempt": 1,
      "node_id": "CR_kwDOHU3a0M8AAAAC_TT1og",
      "head_sha": "5a4d56f44cfe7f5ebcea0c465da0bdc37ed62649",
      "url": "https://api.github.com/repos/canonical/zookeeper-operator/actions/jobs/12838040994",
      "html_url": "https://github.com/canonical/zookeeper-operator/actions/runs/4734403462/jobs/8403276189",
      "status": "completed",
      "conclusion": "success",
      "created_at": "2023-04-18T15:54:43Z",
      "started_at": "2023-04-18T15:55:05Z",
      "completed_at": "2023-04-18T15:56:37Z",
      "name": "Unit tests",
      "steps": [
        {
          "name": "Set up job",
          "status": "completed",
          "conclusion": "success",
          "number": 1,
          "started_at": "2023-04-18T16:55:04.000+01:00",
          "completed_at": "2023-04-18T16:55:05.000+01:00"
        },
```

`canonical/zookeeper-operator` links to the full failing job run
`integration-scaling` links to the specific failing job

##### `feat: add mattermost_alert workflow`
- Currently it's hard to keep track of downstream repos (e.g bundles). We would like a way of notifying Charming teams of CI run failures. Hopefully particularly useful for the cron jobs
- As we use `matrix.tox-environments` for the majority of our integration tests within DPE, we can't reasonably add this workflow as a `Step`, as we'd get N alerts for N failures
- See [screenshots/mattermost_alert.png](https://raw.githubusercontent.com/canonical/data-platform-workflows/7ce248211cdbf02b9bbab52c018121c976923a59/screenshots/mattermost_alert.png) for example output
- Posts made through the Mattermost Channel API using a bot account
    - There's an open IS ticket for the creation of a Mattermost bot named `dataplatformbot`. My hope is that it makes it easier to share the token between our teams.
    - NOTE - Have not yet validated this works, or if the API is public, will do after bot account is created ([open IS ticket](https://portal.admin.canonical.com/C157383)). There are also webhooks to try if that doesn't work but I'm confident it will 

### Avoiding Alert Spam
It's my experience that teams tend to start ignoring alerts if they're too noisy, or they get pings that oftentimes don't involve them.
We can avoid this somewhat by recommending people mute the channel, which will avoid audible 'ping's, but still give a red `(1)` icon to refer to. Muting isn't a great solution however, just a thought for now.

### What workflows to run on?
I suggest we don't run this on PR to `main`, as it would be unbearably noisy. Given our current workflows, I think it makes sense to be a part of our `Publish` jobs, and (ideally) on our `Tests` jobs when ran using `schedule: - cron`.
As we run both on our `Tests` jobs, we'd have to only run when scheduled. I don't yet have the token to validate but I think this can be solved with `${{ github.event_name == 'schedule' }}` and will add.

### Extensibility for other charming teams
In the example below, I reference a channel created called `data-platform-alerts`, which we could use if we want. If teams want their own private channel (`sql-alerts`, `bigdata-alerts`, `cos-alerts` etc), they can do that using `channel-id`.
If they only want to mention a few specific individuals (as teams grow), they can do that using `mention-users`. If they want to add alerting to our public channels (not sure why yet but it's possible), users can do that using `server`, but defaults to internal MM.

### Announcement
Will write a Discourse post if appropriate once merged and validated.

### Cleanup
If implemented, the channels will be a massive list of spam. It's possible in the [MM API](https://api.mattermost.com/#tag/posts) to clean up posts. We could consider this as a 'success' condition (as opposed to the 'failure' only right now).

### CODE REVIEW

**Review by delgod — APPROVED** [review]

File: .github/workflows/mattermost_alert.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+### Basic Example
+```yaml
+notify-failure:
+  name: Notify Failure to Mattermost
+  runs-on: ubuntu-latest
+  needs:
+    - integration-test  # whatever jobs you'd like to ensure runs first
+  steps:
+    - name: Notify Failure To Mattermost
+      uses: canonical/data-platform-workflows/.github/workflows/mattermost_alert.yaml
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply .github/workflows/mattermost_alert.md:16]: I suspect there'll be issues there tbh, but I can't run it until I get the bot creds, will fix that once I can actually run and check!

**Review by taurus-forever — COMMENTED** [review]
1) I like integrations!
2) Are we reinventing: https://github.com/mattermost/mattermost-plugin-github#about-the-github-plugin ?

**Review by deusebio — APPROVED** [review]
LGTM!

One thought I had was if one could make it more parametric when shooting the message to mattermost, instead of hard-coding that [here](https://github.com/canonical/data-platform-workflows/pull/45/files#diff-172cfdf6058805b5f75962527416a36dd737c54c9a8ca46d8ab293d7888ca06dR50). But as a first go, it looks very good and the customization of specifying the condition could definitely be done at a later stage, if / when needed

File: .github/workflows/mattermost_alert.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (taurus-forever) [reply .github/workflows/mattermost_alert.md:16]: @marcoppenheimer JFMI, are you planning to finish this? No rush/pressure, just curiosity. Tnx!

**Pull request closed without merging**